src/main/java/com/limechain/trie/structure/TrieStructure.java

```diff
@@ -1,30 +1,480 @@
 
+import java.util.Collections;
 import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class TrieStructure<T> {
+    /**
+     * The actual container for the trie's nodes.
+     * Using a Slab guarantees consistency of node indices without the need for internal management.
+     */
+    Slab<TrieNode<T>> nodes;
+
+    /**
+     * Index of the root node within {@link TrieStructure#nodes}.
+     * Null if the trie is empty.
+     */
+    @Nullable
+    Integer rootIndex;
+
+    public TrieStructure() {
+        this.nodes = new Slab<>();
+        this.rootIndex = null;
+    }
+
+    public TrieStructure(int initialCapacity) {
+        this.nodes = new Slab<>(initialCapacity);
+        this.rootIndex = null;
+    }
+
+    /**
+     * @return true if the trie is empty
+     */
+    public boolean isEmpty() {
+        return this.nodes.isEmpty();
+    }
+
+    /**
+     * @return the number of nodes in the trie structure, both
+     *         branch (purely structural, with no storage value) and
+     *         storage nodes (leafs or branches).
+     */
+    public int size() {
+        return this.nodes.size();
+    }
+
+    public Optional<NodeHandle<T>> getRootNode() {
+        if (this.rootIndex == null) {
+            return Optional.empty();
+        }
+
+        return Optional.of(this.nodeHandleAtIndexInner(this.rootIndex));
+    }
 
-public class TrieStructure<T> implements Iterable<TrieNodeIndex> {
 
-    // INTERFACE
     @NotNull
-    @Override
-    public Iterator<TrieNodeIndex> iterator() {
-        //TODO: implement
-        throw new NotImplementedException("Not yet implemented.");
+    NodeHandle<T> nodeHandleAtIndexInner(int nodeIndex) {
+        TrieNode<T> node = this.getNodeAtIndexInner(nodeIndex);
+        return NodeHandle.<T>getConstructor(node.hasStorageValue).apply(this, nodeIndex);
+    }
+
+    //NOTE: maybe rename to `nodeAtKey` or something similar... to align with `getNodeAtIndex`
+    public Entry<T> node(Nibbles key) {
+        return switch (this.existingNodeInner(key)) {
+            case ExistingNodeInnerResult.Found found ->
+                NodeHandle.<T>getConstructor(found.hasStorageValue()).apply(this, found.nodeIndex());
+            case ExistingNodeInnerResult.NotFound notFound -> {
+                Integer closestAncestorIndex = null;
+
+                if (notFound.closestAncestor() != null) {
+                    closestAncestorIndex = notFound.closestAncestor().index();
+                }
+
+                yield new Vacant<>(this, key, closestAncestorIndex);
+            }
+        };
+    }
+    // NOTE:
+    //  Those (six) iteration methods could be reduced, or at least live outside this class,
+    //  but I find them all reasonably useful and couldn't decide for anything particularly cleaner.
+
```

Comment:
This note is rather significant: I'm open for suggestions. I still think this abundance of iteration methods is really useful, but it clutters the public interface with repetitive methods.

Reply:
2 things:

1. Do all of these need to be public? `streamUnordered` is only used privately for example.
2. Is performance the only benefit to the unordered functions? Meaning do we actually have anywhere in the code where we want the elements to be unordered? My guess is that the performance benefit might not be significant at all and if that's the case we could remove all the unordered functions.

Reply:
1. It's for completeness. In principle, we could only leave the `stream_` versions of the methods, but when you want to use the TrieStructure as an iterable in an enhanced for-loop for example, you'd have to do it like:
```java
Iterable<TrieNodeIndex> orderedIndices = () -> trieStructure.streamOrdered().iterator();
for (TrieNodeIndex index : orderedIndices) {

}
```
which is just not convenient. So it's a tradeoff between convenience of usage and cluttering the `TrieStructure`'s body.

2. Honestly, I don't have exact measures on the efficiency gains, but it seems there would be a rather significant difference, since traversing the trie in order is not trivial (and hence, I suspect, a rather expensive operation), although with the same (linear) complexity. 

The main arguments in favour of preserving the unordered iteration is: yes, there'll be a point when we'll use it: - for proof encoding. The reference for this is the current state of Smoldot, which was also the reason these methods exist in the first place. Depends on how far we'd want to stick out the similarities between us and Smoldot. 😄

Reply:
I think it's fine to leave it as it is then. There might be multiple functions but all of them have small bodies so it's not taking that much space.